lib/psych/visitors/yaml_tree.rb

```diff
@@ -310,19 +310,20 @@ def visit_String o
           style = Nodes::Scalar::LITERAL
           plain = false
           quote = false
-        elsif o =~ /\n/
+        elsif o =~ /\n[^\Z]/  # match \n except blank line at the end of string
           style = Nodes::Scalar::LITERAL
         elsif o == '<<'
           style = Nodes::Scalar::SINGLE_QUOTED
           tag   = 'tag:yaml.org,2002:str'
           plain = false
           quote = false
+        elsif o.length > line_width
+          style = Nodes::Scalar::FOLDED
+          o << "\n" unless o =~ /\n\Z/  # to avoid non-default chomping indicator
```

Comment:
>  Why is o assigned to str and then both o and str are used later in this method?

I don't have a good answer for that.  I think we could delete the `str` variable.